### PR TITLE
Polychromatic Riemann Sum

### PR DESCRIPTION
**Pull request by johanmazoyer**
+ This PR introduces the Riemann Sum to do polychromatic images from monochromatic images not necessarily evenly sampled. It introduce a new parameter in the `[modelconfig]` section, `mandatory_wls`. This optional keyword allows the introduction of specific wavelengths that need to appear to simulate the polychromatic images. I've made it a "hidden parameter" which means it's not in the parameter files but if you know what you're doing you can modify this. 
These new wavelengths must be in the range ]wavelength_0 - Delta_wav / 2 , wavelength_0 + Delta_wav / 2[. Default is an empty list ('mandatory_wls =  , '). In this case, the nb_wav simulation wavelengths are chosen to be well equally distributed in the Delta_wav. Following figure is in the docs to understand both case: 

image

We then do a Riemann sum to create the polychromatic image:
image

+ This new option allows more freedom in the choice of the estimation and correction wavelength. This PR introduces a new parameter in `[Estimationconfig]`: `estimation_wls`. This allows us to hand pick the correction and estimation wavelengths that we want, even if they are not well centered or evenly space in the bandwidth. I have checked that photon noise should be working. 

+ I modified THD2 class to automatically include `estimation_wls` in `mandatory_wls`.

+ Lot of changes in the docs to be a better documentation

+ Solve some errors in polychromatic correction. This is now quite well tested and I think this is working and ready for the testbed

+ up to v2.6

Docs for this branch can be found here : https://asterix-hci.readthedocs.io/en/polychrom_rieman_sum/

### CODE REVIEW

File: Asterix/tests/test_wfsc.py
Comment (ivalaginja) [inline Asterix/tests/test_wfsc.py:84]: typos in depend~s~

File: Asterix/tests/test_wfsc.py
Comment (ivalaginja) [inline Asterix/tests/test_wfsc.py:109]: typos in depend~s~

File: Asterix/tests/test_wfsc.py
Comment (ivalaginja) [inline Asterix/tests/test_wfsc.py:110]: typo in draw*n*

File: Asterix/tests/test_wfsc.py
Comment (ivalaginja) [inline Asterix/tests/test_wfsc.py:173]: typos in depend~s~

File: Asterix/tests/test_wfsc.py
Comment (ivalaginja) [inline Asterix/tests/test_wfsc.py:174]: typo in draw*n*

File: Asterix/tests/test_wfsc.py
Comment (ivalaginja) [inline Asterix/tests/test_wfsc.py:140]: typos in depend~s~

File: Asterix/tests/test_wfsc.py
Comment (ivalaginja) [inline Asterix/tests/test_wfsc.py:141]: typo in draw*n*

File: docs/correction.rst
Comment (ivalaginja) [inline docs/correction.rst:155]: application to DM and measurement of focal plane

File: setup.py
Comment (ivalaginja) [inline setup.py:29]: Did you actually test all of these versions? Including a full conda install with these Python versions, from scratch?

**Review by ivalaginja — COMMENTED** [review]
Looks mostly fine to me but still trying to figure out how it works and whether it plays nice with the testbed controls.

File: Asterix/optics/optical_systems.py
Comment (ivalaginja) [inline Asterix/optics/optical_systems.py:126]: Comment by @rgalicher on lines 126+127 :
The science_sampling is the one for the central wavelength but when matrices are saved for "estimation_wl", the sampling value should change in the filename. It seems the value changes for the calculation. This is just a naming thing.

**Comment by rgalicher** [comment]
Need to set automatically the wavelength for estimation when monochromatic Ignoring all useless keyword as estimation_wls nb_wav_estim polychromatic etc

**Comment by ivalaginja** [comment]
We were able to create matrices on this branch today that we could use to run on the testbed. Setup for THD2 in configfile:

Set central wavelength to shortest wavelength on testbed since this will create the smallest DH in terms of pixels, which is then used for all wavelengths:
`wavelength_0 = 637e-9`

Set the wavelength range large enough so that it covers your wavelength furthest away from central wavelength (in our case 637 --> 783 nm):
`Delta_wav = 300e-9`

Set the focal-plane sampling at the central wavelength:
`Science_sampling = 5.9`

Tell the code we want to run in multiple wavelengths:
`polychromatic = 'multiwl'`

Tell the code which wavelengths to include in the estimation since these are the wavelengths we need all matrices for:
`estimation_wls = 637e-9, 705e-9, 783.25e-9`

And in `estimator.py`, do the following change (see #163):
`if self.Estim_sampling < 3:` --> `if self.Estim_sampling <= 2:`

---------

This digs a decent DH in simulations and we manage to set up the testbed with this. For me this means this PR is good to go provided all comments get addressed and we tweak things for the testbed independently.

File: Asterix/Example_param_file.ini
Comment (johanmazoyer) [reply Asterix/Example_param_file.ini:25]: ok

File: Asterix/Example_param_file.ini
Comment (johanmazoyer) [reply Asterix/Example_param_file.ini:243]: better ? 

    # If polychromatic = 'broadband_pwprobes', estimation is done using broadband images using
    # wavelength_0 and Delta_wav parameters and Correction is done at wavelength_0.
    # For other cases, you can use 2 different methods:

File: Asterix/Example_param_file.ini
Comment (johanmazoyer) [reply Asterix/Example_param_file.ini:239]: yes it is

File: Asterix/tests/param_file_tests.ini
Comment (johanmazoyer) [reply Asterix/tests/param_file_tests.ini:244]: ok

File: Asterix/tests/param_file_tests.ini
Comment (johanmazoyer) [reply Asterix/tests/param_file_tests.ini:239]: broadband_pwprobes still exists

**Comment by johanmazoyer** [comment]
> Need to set automatically the wavelength for estimation when monochromatic Ignoring all useless keyword as estimation_wls nb_wav_estim polychromatic etc

I think this is already the case ? Or am I mistaken

File: Asterix/optics/optical_systems.py
Comment (johanmazoyer) [reply Asterix/optics/optical_systems.py:126]: ok done

File: setup.py
Comment (johanmazoyer) [reply setup.py:29]: grblblb, yes I just did.

**Comment by johanmazoyer** [comment]
> We were able to create matrices on this branch today that we could use to run on the testbed. Setup for THD2 in configfile:
> 
> Set central wavelength to shortest wavelength on testbed since this will create the smallest DH in terms of pixels, which is then used for all wavelengths: `wavelength_0 = 637e-9`

Argh did not think of that. ok you tweaked it but mayeb we need to be more careful in our definition. For the moment, the DH is define unsing lambd0/D and the whole simulation part assume that lambda_0 is at the center of the correction BW 
I think it would be more in line with the code to actually modify the DH size than wavelength result, but equivalent for the testbed part

> Set the wavelength range large enough so that it covers your wavelength furthest away from central wavelength (in our case 637 --> 783 nm): `Delta_wav = 300e-9`
> 
> Set the focal-plane sampling at the central wavelength: `Science_sampling = 5.9`
> 
> Tell the code we want to run in multiple wavelengths: `polychromatic = 'multiwl'`
> 
> Tell the code which wavelengths to include in the estimation since these are the wavelengths we need all matrices for: `estimation_wls = 637e-9, 705e-9, 783.25e-9`
> 
> Make sure the number of wavelengths for the estimation is indeed what we give it in the line above (see also suggestion to make this adjustment automatically [here](https://github.com/johanmazoyer/Asterix/pull/159#issuecomment-1449910724)): `nb_wav_estim = 3`
> 

Done
> And in `estimator.py`, do the following change (see #163): `if self.Estim_sampling < 3:` --> `if self.Estim_sampling <= 2:`
> 

Done, I put 2.5. Please study at waht point we decrease performance.


> A truncation of the inverted control matrix at 700 modes seems to work ok but we didn't investigate much further.
> 
> This digs a decent DH in simulations and we manage to set up the testbed with this. For me this means this PR is good to go provided all comments get addressed and we tweak things for the testbed independently.

File: Asterix/Example_param_file.ini
Comment (ivalaginja) [reply Asterix/Example_param_file.ini]: I think so

File: Asterix/wfsc/estimator.py
Comment (ivalaginja) [inline Asterix/wfsc/estimator.py:145]: I think you forgot to fill in the two "parameter" here.

**Review by ivalaginja — COMMENTED** [review]
Looks good to me now. I'd like to take it for another spin on the testbed to make sure everything is fine before approving.

**Review by ivalaginja — DISMISSED** [review]
Just found another issue, when in broadband, the code does not save the results to `Mean_Contrast_DH.fits` properly.

**Comment by johanmazoyer** [comment]
> Just found another issue, when in broadband, the code does not save the results to `Mean_Contrast_DH.fits` properly.

ok, i'll do it in another PR because I kind of loose track of all the changes maid here

**Comment by ivalaginja** [comment]
@johanmazoyer can somebody confirm this works ok on hardware before I approve it? With working ok meaning, the output files are read ok by the Labview controls and it kinda digs a DH, even if not super well right now.

**Review by rgalicher — APPROVED** [review]

**Pull request merged**
